### PR TITLE
[MISC] snake_case concept names again

### PR DESCRIPTION
**Pull request by h-2**
This is step 1 of the renaming and upgrading of range-v3. It is the most disruptive one. What can we do to get this in asap?

@joshuak94 Can you maybe scan over this and have a look if you find something glaringly out of place like `FooBar` being replaced with `bax_bat` instead of of `foo_bar`? I don't think it's feasible to check every line. Also some of the concepts have slightly different names now. Thank you !

@rrahn simd is renamed to simd_concept for now. We can search and replace it later when you and @marehr have agreed on a better name.

### CODE REVIEW

File: include/seqan3/range/container/bitcompressed_vector.hpp [hunk 1]
```diff
@@ -32,9 +32,9 @@ namespace seqan3
 {
 
 /*!\brief A space-optimised version of std::vector that compresses multiple letters into a single byte.
- * \tparam alphabet_type The value type of the container, must satisfy seqan3::Alphabet and not be `&`.
- * \implements seqan3::ReservableContainer
- * \implements seqan3::Cerealisable
+ * \tparam alphabet_type The value type of the container, must satisfy seqan3::alphabet and not be `&`.
+ * \implements seqan3::reservable_container
```
Comment (joshuak94) [inline include/seqan3/range/container/bitcompressed_vector.hpp]: I'm not sure if this matters so much... But reversible is spelled with an I not an A.

File: include/seqan3/range/container/concept.hpp [hunk 2]
```diff
@@ -186,13 +186,13 @@ SEQAN3_CONCEPT SequenceContainer = requires (type val, type val2, type const cva
...
- * \extends std::ranges::RandomAccessRange
- * \brief A more refined container concept than seqan3::SequenceContainer.
+/*!\interface seqan3::random_access_container <>
+ * \extends seqan3::sequence_container
+ * \extends std::ranges::random_access_range
+ * \brief A more refined container concept than seqan3::sequence_container.
  *
  * Adds requirements for `.at()`, `.resize()` and the subscript operator `[]`. Models the subset of the
- * [STL SequenceContainerConcept](http://en.cppreference.com/w/cpp/concept/SequenceContainer) that is supported
+ * [STL sequence_containerConcept](http://en.cppreference.com/w/cpp/concept/sequence_container) that is supported
```
Comment (joshuak94) [inline include/seqan3/range/container/concept.hpp]: Should it just be sequence_container and not sequence_containerConcept?

File: include/seqan3/range/container/dynamic_bitset.hpp [hunk 3]
```diff
@@ -1642,9 +1642,9 @@ class dynamic_bitset
      * \{
      */
     /*!\brief Converts the `dynamic_bitset` to a `std::string`.
-     * \tparam char_t Character type; must model seqan3::Char.
-     * \param[in] zero Character of type char_t representing `false`. Default <code>'0'</code>.
-     * \param[in] one  Character of type char_t representing `true`. Default <code>'1'</code>.
+     * \tparam char_t builtin_characteracter type; must model seqan3::builtin_character.
```
Comment (joshuak94) [inline include/seqan3/range/container/dynamic_bitset.hpp]: Undo these three lines.

File: include/seqan3/range/decorator/gap_decorator.hpp [hunk 4]
```diff
@@ -70,15 +70,15 @@ namespace seqan3
  * ### The seqan3::gap_decorator::iterator type
  *
  * \attention The iterator of the seqan3::gap_decorator does not model the
- *            [Cpp17BidirectionalIterator](https://en.cppreference.com/w/cpp/named_req/BidirectionalIterator)
+ *            [Cpp17bidirectional_iterator](https://en.cppreference.com/w/cpp/named_req/bidirectional_iterator)
```
Comment (joshuak94) [inline include/seqan3/range/decorator/gap_decorator.hpp]: Link doesn't exist anymore.

File: include/seqan3/range/view/deep.hpp [hunk 5]
```diff
@@ -128,12 +128,12 @@ class deep : public detail::adaptor_base<deep<underlying_adaptor_t>, underlying_
 
     /*!\brief Unwrap the internal adaptor closure object and pipe the range into it.
      * \tparam    urng_t                Type of the underlying range.
-     * \tparam    underlying_adaptor_t_ Same as underlying_adaptor_t with possibly different cref qualification.
```
Comment (joshuak94) [inline include/seqan3/range/view/deep.hpp]: Undo

File: doc/howto/write_a_view/solution_view.cpp [hunk 6]
```diff
@@ -49,12 +49,12 @@ class my_iterator : public seqan3::detail::inherited_iterator_base<my_iterator<u
 };
 
 // The inherited_iterator_base creates the necessary code so we also model RandomAccess now!
```
Comment (eseiler) [inline doc/howto/write_a_view/solution_view.cpp]:
```suggestion
// The inherited_iterator_base creates the necessary code so we also model random_access now!
```

?

File: doc/howto/write_an_alphabet/dna2_alphabet.cpp [hunk 7]
```diff
@@ -91,12 +91,12 @@ struct dna2
 //! [writable_alphabet]
 
 //! [writable_alphabet_concept]
-static_assert(seqan3::Alphabet<dna2>);                   // ok
-static_assert(seqan3::WritableAlphabet<dna2>);           // ok
+static_assert(seqan3::alphabet<dna2>);                   // ok
+static_assert(seqan3::writable_alphabet<dna2>);           // ok
```
Comment (eseiler) [inline doc/howto/write_an_alphabet/dna2_alphabet.cpp]: Would be nice to preserve indentation, but it's quite a bit to change...

File: include/seqan3/range/view/pairwise_combine.hpp [hunk 8]
```diff
@@ -288,8 +288,8 @@ class pairwise_combine_view : public ranges::view_interface<pairwise_combine_vie
         //!\}
 
         /*!\name Comparison operators
-         * \brief These operators are available if the `underlying_iterator_type` models std::EqualityComparable or
-         *        std::StrictTotallyOrdered respectively.
+         * \brief These operators are available if the `underlying_iterator_type` models std::equality_comparable or
+         *        std::totally_ordered respectively.
```
Comment (joshuak94) [inline include/seqan3/range/view/pairwise_combine.hpp]: The word strict is gone, is that intentional?

File: include/seqan3/range/view/pairwise_combine.hpp [hunk 9]
```diff
@@ -651,8 +651,8 @@ class pairwise_combine_view : public ranges::view_interface<pairwise_combine_vie
  * \{
  */
 
-//!\brief Deduces the correct template type from a non-view lvalue range by wrapping the range in std::ranges::view::all.
-template <std::ranges::ViewableRange other_range_t>
+//!\brief Deduces the correct template type from a non-view lvalue range by wrapping the range in std::ranges::views::all.
```
Comment (joshuak94) [inline include/seqan3/range/view/pairwise_combine.hpp]: Remove the s

File: include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp [hunk 10]
```diff
@@ -135,8 +135,8 @@ inline bool search_ss(cursor_t cur, query_t & query,
 /*!\brief Searches a query sequence in a bidirectional index using a single search of a search scheme.
  *        Sub-function for searching the remaining part of the current block without any errors.
  * \tparam abort_on_hit     If the flag is set, the search aborts on the first hit.
- * \tparam cursor_t         Must model seqan3::BiFmIndexCursor.
- * \tparam query_t          Must model std::ranges::RandomAccessRange over the index's alphabet.
+ * \tparam cursor_t         Must model seqan3::bi_fm_index_specialisationCursor.
```
Comment (joshuak94) [inline include/seqan3/search/algorithm/detail/search_scheme_algorithm.hpp]: bi_fm_index_specialisation_cursor?

File: include/seqan3/search/algorithm/detail/search_trivial.hpp [hunk 11]
```diff
@@ -30,8 +30,8 @@ namespace seqan3::detail
 
 /*!\brief Searches a query sequence in an index using trivial backtracking.
  * \tparam abort_on_hit  If the flag is set, the search algorithm aborts on the first hit.
- * \tparam cursor_t      Must model seqan3::FmIndexCursor.
- * \tparam query_t       Must model std::ranges::InputRange over the index's alphabet.
+ * \tparam cursor_t      Must model seqan3::fm_index_specialisationCursor.
```
Comment (joshuak94) [inline include/seqan3/search/algorithm/detail/search_trivial.hpp]: Here too

File: include/seqan3/std/concepts [hunk 12]
```diff
@@ -435,167 +435,167 @@ SEQAN3_CONCEPT EqualityComparableWith =
...
-    Predicate<R, T, U> &&
-    Predicate<R, U, T>;
+    predicate<R, T, U> &&
+    predicate<R, U, T>;
 //!\endcond
 
-/*!\interface   std::StrictWeakOrder <>
- * \extends     std::Relation
- * \brief       The concept StrictWeakOrder<R, T, U> specifies that the Relation R imposes a strict weak ordering
+/*!\interface   std::strict_weak_order <>
```
Comment (joshuak94) [inline include/seqan3/std/concepts]: Why does strict stay in this concept name, and not in StrictTotallyOrdered and others?

File: include/seqan3/std/concepts [hunk 13]
```diff
@@ -65,112 +65,112 @@ namespace std
 //  Core language concepts
 // ============================================================================
 
-/*!\interface std::Same <>
- * \brief The concept `std::Same<T, U>` is satisfied if and only if T and U denote the same type.
- * \sa https://en.cppreference.com/w/cpp/concepts/Same
+/*!\interface std::same_as <>
+ * \brief The concept `std::same_as<T, U>` is satisfied if and only if T and U denote the same type.
+ * \sa https://en.cppreference.com/w/cpp/concepts/same_as
```
Comment (joshuak94) [inline include/seqan3/std/concepts]: Generally speaking, should the URLs to cppreference all change as well?

File: include/seqan3/std/ranges [hunk 14]
```diff
@@ -392,21 +392,21 @@ namespace std::ranges::view
  */
 inline constexpr auto all = ::ranges::view::all;
 
-} // namespace std::ranges::view
+} // namespace std::ranges::views
```
Comment (joshuak94) [inline include/seqan3/std/ranges]: here and line 406 why views?

File: include/seqan3/std/ranges [hunk 15]
```diff
@@ -490,7 +490,7 @@ inline constexpr auto drop = ::ranges::view::drop;
  *                      to true (and returns the rest).
  * \tparam urng_t       The type of the range being processed. See below for requirements. [template parameter is
  *                      omitted in pipe notation]
- * \tparam fun_t        The type of the functor; must model std::IndirectUnaryPredicate with
+ * \tparam fun_t        The type of the functor; must model std::IndirectUnarypredicate with
```
Comment (joshuak94) [inline include/seqan3/std/ranges]: This didn't change.

File: include/seqan3/std/ranges [hunk 16]
```diff
@@ -504,25 +504,25 @@ inline constexpr auto drop = ::ranges::view::drop;
...
+ * | std::ranges::output_range        |                                       | *preserved*                                        |
+ * | seqan3::const_iterable_range      |                                       | depends on functor (see below)                     |
  * |                                 |                                       |                                                    |
  * | seqan3::reference_t             |                                       | seqan3::reference_t<urng_t>                        |
  *
  * See the \link view view submodule documentation \endlink for detailed descriptions of the view properties.
  *
  * This view only *preserves* certain concepts if and only if the specified functor also models
- * std::IndirectRegularUnaryInvocable<fun_t, reference_t<urng_t>, i.e.
+ * std::IndirectregularUnaryinvocable<fun_t, reference_t<urng_t>, i.e.
```
Comment (joshuak94) [inline include/seqan3/std/ranges]: Also didn't change.

File: include/seqan3/std/ranges [hunk 17]
```diff
@@ -569,7 +569,7 @@ inline constexpr auto drop_while = ::ranges::view::drop_while;
  */
 inline constexpr auto filter = ::ranges::view::filter;
 
-} // namespace std::ranges::view
+} // namespace std::ranges::views
```
Comment (joshuak94) [inline include/seqan3/std/ranges]: view

File: include/seqan3/std/ranges [hunk 18]
```diff
@@ -612,7 +612,7 @@ using istream_view = ::ranges::istream_view<value_type>;
 
 } // namespace std::ranges
 
-namespace std::ranges::view
+namespace std::ranges::views
```
Comment (joshuak94) [inline include/seqan3/std/ranges]: view

File: include/seqan3/std/ranges [hunk 19]
```diff
@@ -952,12 +952,12 @@ inline constexpr auto transform = ::ranges::view::transform;
  */
 inline constexpr auto zip = ::ranges::view::zip;
 
-} // namespace std::ranges::view
+} // namespace std::ranges::views
```
Comment (joshuak94) [inline include/seqan3/std/ranges]: Here and below, why views?

File: doc/howto/write_a_view/solution_view.cpp (hunk 6 above)
Comment (h-2) [reply doc/howto/write_a_view/solution_view.cpp]: Well it should have been named like the concept :man_shrugging:

File: doc/howto/write_an_alphabet/dna2_alphabet.cpp (hunk 7 above)
Comment (h-2) [reply doc/howto/write_an_alphabet/dna2_alphabet.cpp]: Yes, but I don't think it's possible in a generic way.

File: test/documentation/seqan3.css [hunk 20]
```diff
@@ -63,7 +63,7 @@ hr.footer {
 div.fragment {
     font-size: 12pt;
     line-height: 30px;
-    font-family: "Ubuntu Mono", "SFMono-Regular",source-code-pro,Menlo,Monaco,Consolas,"Roboto Mono","Droid Sans Mono","Liberation Mono",Consolas,"Courier New",Courier,monospace;
+    font-family: "Ubuntu Mono", "SFMono-regular",source-code-pro,Menlo,Monaco,Consolas,"Roboto Mono","Droid Sans Mono","Liberation Mono",Consolas,"Courier New",Courier,monospace;
```
Comment (joshuak94) [inline test/documentation/seqan3.css]: Think this file should be excluded.

File: test/documentation/DoxygenLayout.xml [hunk 21]
```diff
@@ -13,7 +13,7 @@
             <tab type="user"  visible="yes" title="First steps"     url="@ref tutorial_first_example" intro=""/>
             <tab type="user"  visible="yes" title="Parsing command line arguments"       url="@ref tutorial_argument_parser"     intro=""/>
             <tab type="user"  visible="yes" title="C++ Concepts"    url="@ref tutorial_concepts"      intro=""/>
-            <tab type="user"  visible="yes" title="Alphabets"       url="@ref tutorial_alphabets"     intro=""/>
+            <tab type="user"  visible="yes" title="alphabets"       url="@ref tutorial_alphabets"     intro=""/>
```
Comment (joshuak94) [inline test/documentation/DoxygenLayout.xml]: Should this file be excluded?

File: test/performance/io/format_fasta_benchmark.cpp [hunk 22]
```diff
@@ -79,7 +79,7 @@ BENCHMARK(write3);
 void write2(benchmark::State & state)
 {
     std::ostringstream ostream;
-    seqan::CharString id = fasta_hdr;
+    seqan::builtin_characterString id = fasta_hdr;
```
Comment (joshuak94) [inline test/performance/io/format_fasta_benchmark.cpp]: Probably char_string or builtin_character_string right?

File: test/performance/io/stream_output_benchmark.cpp [hunk 23]
```diff
@@ -141,7 +141,7 @@ void seqan2_compressed(benchmark::State & state)
 {
     std::ostringstream os;
 
-    seqan::VirtualStream<char, seqan::Output> ogzf;
+    seqan::Virtualstream_REMOVEME<char, seqan::Output> ogzf;
```
Comment (joshuak94) [inline test/performance/io/stream_output_benchmark.cpp]: virtual_stream

File: doc/howto/write_an_alphabet/dna2_alphabet.cpp (hunk 7 above)
Comment (h-2) [reply doc/howto/write_an_alphabet/dna2_alphabet.cpp]: I mean, if you want to to go over the diff and correct for spacing with committable suggestions, I will of course integrate them. But it's a large diff.

**Comment by h-2** [comment]
> Do you have the renaming list somewhere?

I do, but it took almost four hours to fix things after applying the rename script so I don't want to run it again.

In general it's not ideal, but I would suggest fixing some things after this has been merged. For example all lines that contain views will be touched when renaming the views namespace.

File: test/seqan3-test.cmake [hunk 24]
```diff
@@ -80,7 +80,7 @@ target_link_libraries (seqan3_test_header INTERFACE "seqan3::test::performance")
 add_library (seqan3::test::header ALIAS seqan3_test_header)
 
 # ----------------------------------------------------------------------------
-# Commonly shared options for external projects.
+# common_withly shared options for external projects.
```
Comment (joshuak94) [inline test/seqan3-test.cmake]: Exclude this file

File: test/snippet/search/kmer_index/shape.cpp [hunk 25]
```diff
@@ -14,7 +14,7 @@ int main()
     seqan3::debug_stream << s1 << '\n';                    // prints "[1,0,1]"
     seqan3::debug_stream << std::ranges::size(s1) << '\n'; // prints "3"
 
-    seqan3::shape s2{0b101_shape};                         // Same as previous
+    seqan3::shape s2{0b101_shape};                         // same_as as previous
```
Comment (joshuak94) [inline test/snippet/search/kmer_index/shape.cpp]: Undo

File: test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp [hunk 26]
```diff
@@ -181,7 +181,7 @@ TYPED_TEST_CASE(iterator_fixture_two_dimensional, testing_types);
 
 TYPED_TEST(iterator_fixture_two_dimensional, two_dimensional_concept)
 {
-    EXPECT_TRUE(TwoDimensionalMatrixIterator<typename TestFixture::iterator_type>);
+    EXPECT_TRUE(TwoDimensionalmatrixIterator<typename TestFixture::iterator_type>);
```
Comment (joshuak94) [inline test/unit/alignment/matrix/detail/two_dimensional_matrix_test.cpp]: Didn't change

File: test/unit/core/char_operations/char_predicate_test.cpp [hunk 27]
```diff
@@ -49,25 +49,25 @@ TEST(char_predicate, char_predicate_msg)
     EXPECT_EQ(foo<'o'>::msg.str(), "foo_o"s);
 }
 
-TEST(char_predicate, CharPredicate)
+TEST(char_predicate, builtin_characterpredicate)
```
Comment (joshuak94) [inline test/unit/core/char_operations/char_predicate_test.cpp]: Should this be something like builtin_character_predicate

File: test/unit/search/search_scheme_algorithm_test.cpp [hunk 28]
```diff
@@ -154,7 +154,7 @@ inline void test_search_hamming(auto it, text_t const & text, auto const & searc
                      << "Query: " << query << '\n'
                      << "Errors: " << total << ", " << substitution << '\n'
                      << "SS hits: " << hits_ss << '\n'
-                     << "Trivial hits: " << hits_trivial << '\n';
```
Comment (joshuak94) [inline test/unit/search/search_scheme_algorithm_test.cpp]: Undo this and below

**Comment by h-2** [comment]
Thanks for looking into this Josh! It would have been very helpful to have change suggestions, though, then I could have easily fixed all of the comments ...

**Comment by joshuak94** [comment]
Ahhh I'm really sorry about that... I never actually use the "commit suggestion" button, so it never occurred to me that others might...

**Comment by h-2** [comment]
Yeah, I understand that. I usually don't use it either, but in this case it would have been quite helpful. I will just go through it, no problem!

File: include/seqan3/alphabet/composite/detail.hpp [hunk 29]
```diff
@@ -136,18 +136,18 @@ struct weakly_equality_comparable_with
...
  * Returns an std::true_type if the `type` is comparable via <,<=,>,>= to `T`.
  */
 template <typename T>
-struct weakly_ordered_with
+struct weakly_ordered_with_
 {
     //!\brief The returned type when invoked.
     template <typename type>
-    using invoke = std::integral_constant<bool, WeaklyOrderedWith<type, T>>;
+    using invoke = std::integral_constant<bool, weakly_ordered_with<type, T>>;
```
[the remark this answers is not in the document]
Comment (h-2) [reply include/seqan3/alphabet/composite/detail.hpp]: It all makes sense, but is non-trivial to explain. Short version: now that we renamed concepts they sometimes conflict with struct or parameters defined elsewhere. So the other one get a slight rename now. (they are usually not public so appending _ is ok)

[the remark this answers is not in the document]
Comment (joshuak94) [reply include/seqan3/alphabet/composite/detail.hpp]: Ahhh okay that makes sense.

File: include/seqan3/io/alignment_file/format_bam.hpp [hunk 30]
```diff
@@ -99,17 +99,17 @@ class alignment_file_input_format<format_bam> : alignment_file_input_format<form
     /*!\name Constructors, destructor and assignment
      * \{
      */
-    alignment_file_input_format()                                                noexcept = default; //!< Defaulted.
+    alignment_file_input_format_REMOVEME()                                                noexcept = default; //!< Defaulted.
```
[the remark this answers is not in the document]
Comment (h-2) [reply include/seqan3/io/alignment_file/format_bam.hpp]: Yes, here indeed.

File: include/seqan3/io/alignment_file/input.hpp [hunk 31]
```diff
@@ -377,8 +377,8 @@ template <
                                                                               field::EVALUE,
                                                                               field::BIT_SCORE,
                                                                               field::HEADER_PTR>,
-    detail::TypeListOfAlignmentFileInputFormats  valid_formats_    = type_list<format_sam, format_bam>,
-    std::Integral                                stream_char_type_ = char>
+    detail::type_list_specialisationOfalignment_file_input_formats  valid_formats_    = type_list<format_sam, format_bam>,
```
[the remark this answers is not in the document]
Comment (h-2) [reply include/seqan3/io/alignment_file/input.hpp]: o_O

File: include/seqan3/range/container/bitcompressed_vector.hpp (hunk 1 above)
Comment (h-2) [reply include/seqan3/range/container/bitcompressed_vector.hpp]: Good catch! One of the few things I could never wrap my head around... is there a rule?

File: include/seqan3/range/view/pairwise_combine.hpp (hunk 8 above)
Comment (h-2) [reply include/seqan3/range/view/pairwise_combine.hpp]: yep

File: include/seqan3/std/concepts (hunk 12 above)
Comment (h-2) [reply include/seqan3/std/concepts]: Ask the ISO committee...

File: include/seqan3/std/concepts (hunk 13 above)
Comment (h-2) [reply include/seqan3/std/concepts]: Yes

File: include/seqan3/std/ranges (hunk 14 above)
Comment (h-2) [reply include/seqan3/std/ranges]: hard to explain, but should be this way.

File: include/seqan3/range/container/bitcompressed_vector.hpp (hunk 1 above)
Comment (joshuak94) [reply include/seqan3/range/container/bitcompressed_vector.hpp]: I just googled it... There isn't really a catch all rule haha. I think it's just based on feeling?

**Comment by h-2** [comment]
Thank you for the thorough review, @joshuak94! You don't have to look at it again. If CI passes, we will just merge this and fix remaining things in master later.

**Comment by h-2** [comment]
(I didnt reply to most of the comments, but most of them were correct!)

**Comment by joshuak94** [comment]
Sounds good!

**Pull request merged**